[2.9] expose Merkle Proofs

samoht opened the pull request.

samoht commented on src/irmin/tree_intf.ml:
```diff
@@ -385,5 +386,16 @@ module type Tree = sig
+    (** {2 Proofs} *)
+
+    type proof =
+      [ `Tree of P.Node.Val.proof * (step * proof) list
```
we should probably inline node proofs here, as we don't need to store the step twice and intermediate hashes.

samoht commented:
This is the first draft for #928 which simply expose the underlying shallow inodes. This needs more tests and a better tree proof (the inode proofs are good enough for a first draft I think).

samoht commented:
I've added tests and fixed a few issues. I think this is ready to review.

Will forward-port to 3.0 once merged (as we can release a 2.9 just with this more easily).

samoht commented:
(and now with a fix for irmin's tree tests)

samoht commented:
I've run a few simulations of Merkle proof size with a context of 100k entries to compare between various inode configurations.

Size of Merkle proof for 1 operations:
- Merkle B-trees (32 children): 2kB
- binary Merkle trees         : 0kB

Size of Merkle proof for 100 operations:
- Merkle B-trees (32 children): 98kB
- binary Merkle trees         : 32kB

Size of Merkle proof for 1000 operations:
- Merkle B-trees (32 children): 656kB
- binary Merkle trees         : 248kB

Size of Merkle proof for 10000 operations:
- Merkle B-trees (32 children): 2569kB
- binary Merkle trees         : 1614kB

icristescu commented on src/irmin-pack/inode.ml:
```diff
@@ -671,36 +697,43 @@ struct
-                let t = { hash = lazy pointer; stable = false; v } in
-                entries.(index) <- Some (Ptr.of_target Total t))
+                match aux (depth + 1) tree with
+                | None -> entries.(index) <- Some (Broken pointer)
```
This test seems to fail https://github.com/icristescu/irmin/commit/f4eea14eaeefebf4047b6c3a74ecec06088b0522#diff-dbadcd7de0013fa46dca5ed344f5e516f1e5ad6fb3d05e969834a0c866b18ca9R231 I think because we set `Broken` whenever the hash is `Blinded`?

samoht replied:
yiikes - there is no way to recover the "stable" hash if we only have a partial proof as we need all the values to compute the hash. Will fix this - but we definitively need flattening to have shorter proofs!

samoht replied:
Should be fixed by 775ab10

samoht replied:
Actually I have something simpler -- I'll force push shortly.

samoht replied:
Now in 083b118

samoht commented:
Now with proof extenders (to compress common prefixes).

The result is not great as the proofs are larger (and seemt to take longer to compute):

```
Size of Merkle proof for 1 operations:
- Merkle B-trees (32 children): 2kB
- binary Merkle trees         : 0kB
Size of Merkle proof for 100 operations:
- Merkle B-trees (32 children): 101kB
- binary Merkle trees         : 34kB
Size of Merkle proof for 1000 operations:
- Merkle B-trees (32 children): 675kB
- binary Merkle trees         : 259kB
Size of Merkle proof for 10000 operations:
- Merkle B-trees (32 children): 2601kB
- binary Merkle trees         : 1669kB
```

This is because every index now has to be boxed in a singleton list ; and that the initial store is well balanced.

CraigFe commented:
@samoht: some initial suggestions for changes [here](https://github.com/samoht/irmin/pull/12). I'll rebase it over your latest commit now.

Ngoguey42 commented on src/irmin-pack/inode.ml:
```diff
@@ -961,7 +995,9 @@ struct
     let check_stable layout t =
-      let target_of_ptr = Ptr.target ~cache:true layout in
+      let target_of_ptr =
+        Ptr.target ~cache:true ~force:true "check_stabke" layout
```
```suggestion
        Ptr.target ~cache:true ~force:true "check_stable" layout
```

Ngoguey42 commented on src/irmin-pack/inode.ml:
```diff
@@ -992,6 +1030,99 @@ struct
+                  let e =
+                    match p with
+                    | Inode { proofs = [ (index, proof) ]; _ } ->
+                        (e.index :: index, proof)
```
Is this extender pattern occurring at all in your tests? It think that the probability it occurs in our design is very close to zero.

samoht replied:
yes I'm testing the extender - I think they are used in ethereum to protect against attacks (where people will try to build a very unbalanced tree by using a lot of computing power) and then using this to creat huge proofs.

Ngoguey42 commented on src/irmin-pack/inode.ml:
```diff
@@ -335,8 +345,10 @@ struct
+      let target :
+          type ptr.
+          cache:bool -> force:bool -> string -> ptr layout -> ptr -> ptr t =
+       fun ~cache ~force context layout ->
```
This `force` trick to compute the merkle proof also includes the inodes that were loaded but not modified, is this intended?

You should maybe raise `Dangling_hash` when `force = true` and `{ target = Lazy_loaded entry }`

samoht replied:
the goal is to never call `find` when building a proof as we want the minimal inode/tree which can be used to verify a sequence of operations later on.

I am not sure to understand your remark (as I am not sure to understand exactly what this inode code is doing 😓 ). I am not sure what's the difference? Normally, you can just "touch" all the keys that the sequence of operation does to get your proof. Are you suggesting we can do something better?

E.g. see https://github.com/samoht/irmin/blob/2.8-merke-proofs/src/irmin/tree.ml#L1880-L1883 on we do this currently. But I'm open to a better way :-)

Ngoguey42 replied:
> the goal is to never call find when building a proof as we want the minimal inode/tree which can be used to verify a sequence of operations later on.

I see now! Thanks

samoht commented:
Rebased and squashed - 78ab45e revert the extender part, not sure if we really need those. I let you decide :-) (the safe side might be to keep them to protect against unbalanced inode trees)

Ngoguey42 commented on src/irmin-pack/inode.ml:
```diff
@@ -583,13 +595,20 @@ struct
+                | `Eq x, `Eq y -> `Eq (x + y)
+                | (`Eq x | `Le x), (`Eq y | `Le y) -> `Le (x + y))
+              (`Eq 0) t.pointers
+        | Blinded -> `Le 0
```
Is this `Le 0` or `Ge 0`?

CraigFe replied:
Or even `Ge 1`?

Ngoguey42 replied:
Could be the empty inode

samoht replied:
Fixed in 548e9f6

Ngoguey42 commented on src/irmin-pack/inode.ml:
```diff
@@ -992,6 +1030,78 @@ struct
+        let c = concrete_of_proof 0 proof in
+        of_concrete_exn c
+
+      let of_concrete t = proof_of_concrete (lazy (failwith "blinded root")) t
```
This function doesn't check for stability as `to_proof` does, I think that this is a problem

Ngoguey42 reviewed (commented):
A review on the inodes

CraigFe commented:
Let's re-target this to a `2.9` branch before merging, as our next release is probably `2.9.0` rather than `2.8.1`.

samoht commented on test/irmin-pack/test_tree.ml:
```diff
@@ -144,6 +171,122 @@ let test_fold_undefined () =
+
+let test_proofs steps =
+  (* init the store *)
+  let bindings = bindings steps in
```
might be worth testing this on more complex stores (deeper + a combination of small&large directories) before merging.

samoht commented on test/irmin-pack/test_tree.ml:
```diff
@@ -144,6 +171,122 @@ let test_fold_undefined () =
+      Alcotest.(check_repr Store.Hash.t)
+        (Fmt.str "same hash del %a" Fmt.(Dump.list string) k)
+        (Tree.hash tree) (Tree.hash proof);
+      (tree, proof)
```
we also need an equivalence test for reads

Ngoguey42 commented:
This is ready for merge into 2.9

icristescu commented:
Thanks!
